src/main/java/org/truffleruby/parser/parser/ParserSupport.java

```diff
@@ -987,6 +989,38 @@ public WhenParseNode newWhenNode(SourceIndexLength position, ParseNode expressio
         return new WhenParseNode(position, expressionNodes, bodyNode, nextCase);
     }
 
+    /** Ok I admit that this is somewhat ugly. We post-process a chain of in nodes and analyze them to re-insert them
+     * back into our new CaseInParseNode the way we want. The grammar is being difficult and until I go back into the
+     * depths of that this is where things are.
+     *
+     * @param expression of the case node (e.g. case foo)
+     * @param firstInNode first in (which could also be the else)
+     * @return a new case node */
```

Comment:
For what it's worth, I don't think this is ugly at all. It keeps the grammar clean and all the logic in a single place.

The alternative would be to use the pattern used with `ArrayParser` in the grammar with left-recursion, so something roughly like this:

```
p_case_body     : keyword_in args then compstmt {
                    $$ = new SomeKindOfCaseBodyThing();
                    // or $$ = support.storedCaseInNode;
                    $$.add(support.newInNode($1, $2, $4));
                }
                | p_case_body keyword_in args then compstmt {
                    $$ = $1.add(support.newInNode($2, $3, $5));
                }
                | p_case_body keyword_else compstmt {
                    $$ = $1.set_else($3);
                }
```

Is this problematic with the grammar?

Reply:
Note that this comment was copied from `newCaseNode()`, which was @enebo’s work in d7361b88d4e224e6837111e9e111342ca9581406.

Reply:
Okay! Probably worth changing that later if we decide that's not how we feel about it.